### ISSUE TITLE
[nvidia] Screenshot feature seems to be broken

### ISSUE THREAD

**Issue by Poly-cryptic** [hyprwm/hyprlock#59]
The screenshot feature seems to broken for me and instead of a screenshot, a gray background pops up. These errors are probably related.

> [LOG] [gl] GL_INVALID_OPERATION error generated. <image> and <target> are incompatible
> [LOG] [gl] GL_INVALID_OPERATION error generated. <image> and <target> are incompatible
> [LOG] [gl] GL_INVALID_OPERATION error generated. <image> and <target> are incompatible

Here is my config:

> background {
>     monitor =
>     path = screenshot
> }

I have attached a full verbose log.

[log.txt](https://github.com/hyprwm/hyprlock/files/14366652/log.txt)

**Comment by vaxerski** [comment]
nvidia?

**Comment by Poly-cryptic** [comment]
Yeah, NVidia, I don't know how I forgot that crucial detail.

**Comment by vaxerski** [comment]
https://github.com/ValveSoftware/gamescope/issues/708#issuecomment-1436051840

I am unsure how to check if a modifier is marked as such, I've asked on IRC. Will try to figure something out.

**Comment by vaxerski** [comment]
[patch4.txt](https://github.com/hyprwm/hyprlock/files/14367145/patch4.txt)
hey hey try this

**Comment by Poly-cryptic** [comment]
It doesn't seem to change anything whatsoever.

[log-patch1.txt](https://github.com/hyprwm/hyprlock/files/14367212/log-patch1.txt)

**Comment by vaxerski** [comment]
:(

**Comment by vaxerski** [comment]
[patch69.txt](https://github.com/hyprwm/hyprlock/files/14367344/patch69.txt)
try this :D

**Comment by Poly-cryptic** [comment]
Nothing changed again sadly...

[log-patch2.txt](https://github.com/hyprwm/hyprlock/files/14367416/log-patch2.txt)

**Comment by vaxerski** [comment]
[patch69.txt](https://github.com/hyprwm/hyprlock/files/14367749/patch69.txt)
wont fix anything but add a few logs

**Comment by Poly-cryptic** [comment]
[log-patch3.txt](https://github.com/hyprwm/hyprlock/files/14367808/log-patch3.txt)

**Comment by vaxerski** [comment]
are you on 10bit?

**Comment by Poly-cryptic** [comment]
To be honest, I don't know. How should I go about checking that?

**Comment by vaxerski** [comment]
if you didnt explicitly set 10 bit you aren't running it. you can enable it by adding `,bitdepth,10` to your monitor configs in hyprland.

**Comment by Poly-cryptic** [comment]
Yeah I don't have that enabled and enabling it didn't change anything.

**Comment by vaxerski** [comment]
:(

All issues I see regarding this error is that it happens because nvidia doesn't support importing buffers with linear modifiers, but we filter those unsupported mods and gbm chooses a non-linear one :(

I dont know honestly

**Comment by vaxerski** [comment]
I've committed the patch regardless as it's the correct thing to do anyways. Maybe I'll look into this more tomorrow, I've asked on IRC but it's kinda late and simon zeni doesnt know wtf is up and neither do I

**Comment by Poly-cryptic** [comment]
Yeah, thank you for the quick support. This is why I choose Hyprland.

**Comment by NoisyFlake** [comment]
Same issue, using nvidia-dkms 535 driver.

**Comment by nosduco** [comment]
Also seeing this using nvidia-dkms 550 driver

**Comment by cmdham** [comment]
same issue using nvidia 550 driver, heres my hacky solution:

first set the path to a temp location
``` js
background {
path = /tmp/screenshot.png
}
```
then, every time you run hyprlock, take a screenshot of the monitor and save it to this  location

``` zsh
grim -o DP-1 /tmp/screenshot.png && hyprlock
```

et voilà

**Comment by cxOrz** [comment]
why not use a more compatible way to implement this feature, to capture the screenshot, then save to some temp directory, and use the screenshot.

**Comment by vaxerski** [comment]
because it's slow and wasteful. We can get the screenshot to a cpu buffer, it's just slow.

**Comment by cxOrz** [comment]
I've tried hyprpicker just now, seems it can grab a "screenshot", then pick color directly on the screen. Works on my nvidia card. Is it possible to refer to this implementation? :smile:

**Comment by vaxerski** [comment]
hyprpicker is not gpu-accelerated, so it grabs a cpu buffer, exactly what I said, it's slow.

**Comment by samos667** [comment]
> same issue using nvidia 550 driver, heres my hacky solution:
> 
> first set the path to a temp location
> 
> ```js
> background {
> path = /tmp/screenshot.png
> }
> ```
> 
> then, every time you run hyprlock, take a screenshot of the monitor and save it to this location
> 
> ```shell
> grim -o DP-1 /tmp/screenshot.png && hyprlock
> ```
> 
> et voilà

You can use "-l 0" as option of grim command to not compress screenshot and make command execution 4-5x more fast:
![image](https://github.com/user-attachments/assets/458d9206-57f5-4097-832e-98eb5c7cd097)

This makes a big difference when using multi-screen and/or high-resolution displays.

btw f*ck nv*dia everyday

**Comment by samos667** [comment]
For multiscreen, you can also use the screenshot commands in parallel. This will speed things up a bit too:
```bash
#!/usr/bin/env bash

grim -o HDMI-A-1 -l 0 /tmp/hyprlock_screenshot1.png & # run this command in background
grim -o DP-1 -l 0 /tmp/hyprlock_screenshot2.png & # idem
wait && # wait background commands to finish
hyprlock # so hyprlock will only start when screenshot(s) are done
```

**Comment by samos667** [comment]
@vaxerski 

It's a silly (and certainly naive) question, but in the future, with the open kernel module https://github.com/NVIDIA/open-gpu-kernel-modules
is there a chance that NVIDIA hardware will get better support in the Linux ecosystem ? Or have the red team won once and for all ?

**Comment by Arroquw** [comment]
Why not put us nvidia users up with a slow (but working) way until nvidia finally decides to do something about this?

**Comment by vaxerski** [comment]
because nobody can be arsed to code it, it seems

**Comment by TotalTaxAmount** [comment]
Any update on this?

**Comment by JeffDess** [comment]
Same thing is happening to me on 565 driver. I am getting a similar log than what originally reported (edited to include only one screen):

```log 
[TRACE] [sc] wlr frame copied
[TRACE] timer thread firing
[TRACE] [core] got poll event
[TRACE] got wl event
[LOG] Got fractional scale: 1
[LOG] configure with serial 527
[LOG] Configuring surface for logical [Vector2D: x: 3440, y: 1440] and pixel [Vector2D: x: 3440, y: 1440]
[TRACE] render lock
[TRACE] frame 3
[LOG] Got fractional scale: 1
[TRACE] [sc] wlrOnReady for 0x5d20f0fa5dd8
[LOG] [gl] GL_INVALID_OPERATION error generated. <image> and <target> are incompatible
[LOG] Got dma frame with size [Vector2D: x: 3440, y: 1440]
```

I've implemented the workaround for 3 monitors, I honestly cannot tell a performance difference with my other AMD GPU machine. Probably a few ms if I were to measure it, but nothing I would notice.

The only thing that bothers me is the screenshots left on /tmp (reminds me of a certain Recall feature 😆). 

So I delete them on unlock:

```bash
#!/bin/bash

grim -o HDMI-A-1 -l 0 /tmp/hyprlock_hdmi1.png &
grim -o DP-1 -l 0 /tmp/hyprlock_dp1.png &
grim -o DP-2 -l 0 /tmp/hyprlock_dp2.png &
hyprlock && rm -f /tmp/hyprlock_*.png
```

**Comment by alttabber** [comment]
> Same thing is happening to me on 565 driver. I am getting a similar log than what originally reported (edited to include only one screen):
> 
> ```
> [TRACE] [sc] wlr frame copied
> [TRACE] timer thread firing
> [TRACE] [core] got poll event
> [TRACE] got wl event
> [LOG] Got fractional scale: 1
> [LOG] configure with serial 527
> [LOG] Configuring surface for logical [Vector2D: x: 3440, y: 1440] and pixel [Vector2D: x: 3440, y: 1440]
> [TRACE] render lock
> [TRACE] frame 3
> [LOG] Got fractional scale: 1
> [TRACE] [sc] wlrOnReady for 0x5d20f0fa5dd8
> [LOG] [gl] GL_INVALID_OPERATION error generated. <image> and <target> are incompatible
> [LOG] Got dma frame with size [Vector2D: x: 3440, y: 1440]
> ```
> 
> I've implemented the workaround for 3 monitors, I honestly cannot tell a performance difference with my other AMD GPU machine. Probably a few ms if I were to measure it, but nothing I would notice.
> 
> The only thing that bothers me is the screenshots left on /tmp (reminds me of a certain Recall feature 😆).
> 
> So I delete them on unlock:
> 
> ```shell
> #!/bin/bash
> 
> grim -o HDMI-A-1 -l 0 /tmp/hyprlock_hdmi1.png &
> grim -o DP-1 -l 0 /tmp/hyprlock_dp1.png &
> grim -o DP-2 -l 0 /tmp/hyprlock_dp2.png &
> hyprlock && rm -f /tmp/hyprlock_*.png
> ```

It does look like an improvement, but I don't think this issue will be fixed by Nvidia's side: Nvidia pretty much told compositors to focus on Vulkan.

I have a feeling that if Hyprlock gets a Vulkan rewrite this issue will magically fix itself, but I don't think anyone actually wants to do it.
